File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int totalGoodNodes = 0;
    void goodNodes_01(TreeNode* node,int maxVal) {
        //Base Case
        if(!node) return;
        
        //Check if Good Node or Not
        if(node->val >= maxVal) {
            totalGoodNodes++;
            maxVal = node->val;
        }
        
        //Recursive Call
        goodNodes_01(node->left,maxVal);
        goodNodes_01(node->right,maxVal);
    }
    
    int goodNodes(TreeNode* root) {
        goodNodes_01(root,-(int)1e5);
        return totalGoodNodes;
    }
};

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int totalGoodNodes = 0;
    public void goodNodes_01(TreeNode node,int maxVal) {
        //Base Case
        if(node==null)return;
        // Checking if current node is a good node or not
        if(node.val >= maxVal) {
            totalGoodNodes++;
            maxVal = node.val;
        }
        
        //Left and right recursive call for traversing the tree.
        goodNodes_01(node.left,maxVal);
        goodNodes_01(node.right,maxVal);
        
    }
    
    public int goodNodes(TreeNode root) {
        goodNodes_01(root,-(int)1e5);
        return totalGoodNodes;
    }
}